cyto_dl/models/im2im/diffusion_autoencoder.py

```diff
@@ -0,0 +1,227 @@
+        timesteps = torch.randint(
+            0,
+            self.inferer.scheduler.num_train_timesteps,
+            (x_diff.shape[0],),
+            device=x_diff.device,
+            dtype=torch.long,
+        )
+        # latent is B x C x 1
+        latent = self.semantic_encoder(x_cond).unsqueeze(2)
+        noise_pred = self.inferer(
+            inputs=x_diff,
+            diffusion_model=self.autoencoder,
+            noise=noise,
+            timesteps=timesteps,
+            condition=latent,
+        )
+        return noise, noise_pred, latent
+
+    def _generate_image(self, noise, cond):
+        with torch.no_grad():
+            sample, intermediates = self.inferer.sample(
+                input_noise=noise,
+                diffusion_model=self.autoencoder,
+                scheduler=self.scheduler,
+                save_intermediates=True,
+                conditioning=cond,
+                verbose=False,
+                intermediate_steps=1,
+            )
+        # during training, final image is all-nan, try to return last non-nan image
+        if torch.any(torch.isnan(sample)):
+            while torch.any(torch.isnan(sample)) and len(intermediates) > 0:
+                sample = intermediates.pop(-1)
+        return sample
+
+    def generate_image(self, stage, cond_img, diff_img):
+        """Save the sequence of denoising steps."""
+        self.scheduler.set_timesteps(num_inference_steps=100)
+        with torch.no_grad():
+            cond = self.semantic_encoder(cond_img).unsqueeze(2)
+        noise = torch.randn_like(cond_img, device=self.device)
+        sample = self._generate_image(noise, cond)
+
+        for img, name in zip([cond_img, diff_img, sample], ["cond", "diff", "recon"]):
+            OmeTiffWriter.save(
+                uri=f"{self.hparams.save_dir}/{self.trainer.current_epoch}_{stage}_{name}.tiff",
+                data=detach(img).astype(float),
+            )
+
+    def _to_tensor(self, batch):
+        """convert monai metatensors to tensors."""
+        for k, v in batch.items():
+            if isinstance(v, MetaTensor):
+                batch[k] = v.as_tensor()
+        return batch
+
+    def model_step(self, stage, batch, batch_idx):
+        batch = self._to_tensor(batch)
+        cond_img = batch[self.hparams.condition_key]
+        diff_img = batch[self.diffusion_key]
+        noise, noise_pred, latent = self.forward(cond_img, diff_img)
+        if (self.trainer.current_epoch % self.hparams.save_images_every_n_epochs) == batch_idx == 0:
+            self.generate_image(stage, cond_img[:1], diff_img[:1])
+        diffusion_loss = self.loss(noise, noise_pred)
+
+        return {"loss": diffusion_loss}, latent, None
+
+    def validation_step(self, batch, batch_idx):
+        loss, preds, targets = self.model_step("val", batch, batch_idx)
+        self.compute_metrics(loss, preds, targets, "val")
+        return loss, preds
+
+    def latent_walk(self, cond, save_name):
```

Comment:
is this also a callback?

Reply:
yes, the callback calls this function. I have it split out so you can generate a sequence of images from a batch of latent variables. I'll rename this to make it more clear.